File: frontend/src/story-ledger/contracts.ts
export type JsonPrimitive = string | number | boolean | null;

export type JsonValue =
  | JsonPrimitive
  | readonly JsonValue[]
  | { readonly [key: string]: JsonValue };

export const STORY_LEDGER_FACT_TYPES = [
  "character_state",
  "relationship_state",
  "storyline_event",
  "foreshadow_event",
  "story_time",
  "knowledge_event",
  "world_state",
  "general_fact",
] as const;

export type StoryLedgerFactType = (typeof STORY_LEDGER_FACT_TYPES)[number];

export const STORY_LEDGER_EFFECTIVE_STATES = [
  "current",
  "historical",
  "superseded",
  "source_invalid",
  "batch_reverted",
] as const;

export type StoryLedgerFactEffectiveState =
  (typeof STORY_LEDGER_EFFECTIVE_STATES)[number];

export const STORY_LEDGER_HEALTH_STATES = ["ok", "conflict", "ambiguous"] as const;
export type StoryLedgerFactHealth = (typeof STORY_LEDGER_HEALTH_STATES)[number];

export type StoryLedgerEntityType =
  | "character"
  | "character_instance"
  | "relationship"
  | "storyline"
  | "foreshadow";

export interface StoryLedgerTimelineContext {
  readonly mode: "none" | "single" | "multiple";
  readonly timeline_id: string | null;
  readonly timeline_name: string | null;
  readonly narrative_cutoff: number | null;
}

export interface StoryLedgerEntityReference {
  readonly entity_type: StoryLedgerEntityType;
  readonly entity_id: string;
  readonly label: string;
  readonly lifecycle_state: string | null;
  readonly reference_missing: boolean;
}

export interface StoryLedgerSourceReference {
  readonly source_document_id: string | null;
  readonly document_title: string | null;
  readonly document_position: number | null;
  readonly source_revision_id: string | null;
  readonly revision_number: number | null;
  readonly revision_is_current: boolean | null;
  readonly source_content_hash: string | null;
  readonly source_start: number | null;
  readonly source_end: number | null;
  readonly binding_state: string | null;
  readonly commit_batch_id: string | null;
  readonly evidence_available: boolean;
}

export interface StoryLedgerFactItem {
  readonly id: string;
  readonly fact_type: string;
  readonly subject: string;
  readonly predicate: string;
  readonly object_preview: string;
  readonly object_truncated: boolean;
  readonly timeline_id: string | null;
  readonly dimension: string | null;
  readonly event_kind: string | null;
  readonly story_sequence: number | null;
  readonly created_at: string;
  readonly effective_state: StoryLedgerFactEffectiveState;
  readonly effective_reason_codes: readonly string[];
  readonly included_in_current_projection: boolean;
  readonly health: StoryLedgerFactHealth;
  readonly health_reason_codes: readonly string[];
  readonly entities: readonly StoryLedgerEntityReference[];
  readonly source: StoryLedgerSourceReference | null;
}

export interface StoryLedgerEventLinkView {
  readonly id: string;
  readonly direction: "incoming" | "outgoing";
  readonly link_type: string;
  readonly other_fact_id: string;
  readonly details: Readonly<Record<string, JsonValue>>;
  readonly created_at: string;
}

export interface StoryLedgerBindingView {
  readonly id: string;
  readonly source_document_id: string;
  readonly source_revision_id: string;
  readonly source_content_hash: string;
  readonly validity_state: string;
  readonly proposal_item_id: string | null;
  readonly commit_batch_id: string | null;
  readonly commit_batch_state: string | null;
  readonly created_at: string;
}

export interface StoryLedgerSummary {
  readonly schema_version: "story-ledger-summary/1";
  readonly novel_id: string;
  readonly ledger_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly filter_sha256: string;
  readonly total: number;
  readonly by_fact_type: Readonly<Record<string, number>>;
  readonly by_effective_state: Readonly<Record<string, number>>;
  readonly by_health: Readonly<Record<string, number>>;
  readonly review_required: number;
}

export interface StoryLedgerFactPage {
  readonly schema_version: "story-ledger-page/1";
  readonly novel_id: string;
  readonly ledger_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly filter_sha256: string;
  readonly items: readonly StoryLedgerFactItem[];
  readonly next_cursor: string | null;
}

export interface StoryLedgerFactDetail {
  readonly schema_version: "story-ledger-fact-detail/1";
  readonly novel_id: string;
  readonly ledger_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly item: StoryLedgerFactItem;
  readonly object_text: string;
  readonly details: Readonly<Record<string, JsonValue>>;
  readonly story_time: Readonly<Record<string, JsonValue>> | null;
  readonly visibility: Readonly<Record<string, JsonValue>> | null;
  readonly lifecycle_status: string;
  readonly schema_version_of_fact: string | null;
  readonly event_fingerprint: string | null;
  readonly event_links: readonly StoryLedgerEventLinkView[];
  readonly bindings: readonly StoryLedgerBindingView[];
}

export interface StoryLedgerSourceExcerpt {
  readonly schema_version: "story-ledger-source/1";
  readonly novel_id: string;
  readonly fact_id: string;
  readonly ledger_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly available: boolean;
  readonly unavailable_reason: string | null;
  readonly document_id: string | null;
  readonly document_title: string | null;
  readonly document_position: number | null;
  readonly revision_id: string | null;
  readonly revision_number: number | null;
  readonly revision_is_current: boolean | null;
  readonly source_content_hash: string | null;
  readonly source_range_hash: string | null;
  readonly source_start: number | null;
  readonly source_end: number | null;
  readonly excerpt: string;
  readonly excerpt_start: number | null;
  readonly excerpt_end: number | null;
  readonly highlight_start: number | null;
  readonly highlight_end: number | null;
  readonly truncated_before: boolean;
  readonly truncated_after: boolean;
}

export interface StoryLedgerFactImpactPreview {
  readonly schema_version: "story-ledger-fact-impact-preview/1";
  readonly novel_id: string;
  readonly fact_id: string;
  readonly preview_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly currently_in_projection: boolean;
  readonly current_projection_fact_count: number;
  readonly related_event_link_count: number;
  readonly embedding_rebuild_required: boolean;
  readonly commit_batch_ids: readonly string[];
  readonly batch_fact_count: number;
  readonly batch_relationship_count: number;
  readonly correction_supported: boolean;
  readonly correction_block_reason: string | null;
}

export interface StoryLedgerBatchFactImpact extends Readonly<Record<string, JsonValue>> {
  readonly id: string;
  readonly disposition: "preserve_followup" | "supersede" | "preserve";
}

export interface StoryLedgerBatchRelationshipImpact
  extends Readonly<Record<string, JsonValue>> {
  readonly id: string;
  readonly disposition: "preserve_root_reproject_visibility";
}

export interface StoryLedgerBatchImpactPreview {
  readonly schema_version: "story-ledger-batch-impact-preview/1";
  readonly novel_id: string;
  readonly batch_id: string;
  readonly preview_snapshot_token: string;
  readonly story_ledger_version: number;
  readonly timeline: StoryLedgerTimelineContext;
  readonly state: string;
  readonly already_reverted: boolean;
  readonly batch_fact_count: number;
  readonly batch_relationship_count: number;
  readonly facts: readonly StoryLedgerBatchFactImpact[];
  readonly relationships: readonly StoryLedgerBatchRelationshipImpact[];
}

export interface StoryLedgerFilters {
  readonly factTypes?: readonly string[];
  readonly effectiveState?: StoryLedgerFactEffectiveState | null;
  readonly health?: StoryLedgerFactHealth | null;
  readonly dimension?: string | null;
  readonly sourceDocumentId?: string | null;
  readonly commitBatchId?: string | null;
  readonly factTimelineId?: string | null;
  readonly entityType?: StoryLedgerEntityType | null;
  readonly entityId?: string | null;
  readonly reviewOnly?: boolean;
}

export interface StoryLedgerReadScope {
  readonly novelId: string;
  readonly timelineId?: string | null;
  readonly narrativeCutoff?: number | null;
  readonly snapshotToken?: string | null;
}

export interface StoryLedgerPageQuery extends StoryLedgerFilters {
  readonly cursor?: string | null;
  readonly limit?: number;
}

export interface StoryFactCorrectionCommandV1 {
  readonly schema_version: "story-fact-correction/1";
  readonly operation_key: string;
  readonly expected_story_ledger_version: number;
  readonly reason: string;
  readonly replacement: Readonly<Record<string, JsonValue>>;
}

export interface StoryFactCorrectionResultV1 {
  readonly replayed: boolean;
  readonly story_ledger_version: number;
  readonly fact: Readonly<Record<string, JsonValue>>;
}

export interface IntelligenceBatchRevertCommandV1 {
  readonly operation_key: string;
  readonly expected_story_ledger_version: number;
  readonly reason?: string | null;
}

export interface IntelligenceBatchRevertResultV1 {
  readonly replayed: boolean;
  readonly story_ledger_version: number;
  readonly batch_id: string;
}

File: backend/story_ledger/contracts.py
"""Public wire contracts for the whole-novel Story Ledger workspace."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


FactEffectiveState = Literal[
    "current",
    "historical",
    "superseded",
    "source_invalid",
    "batch_reverted",
]
FactHealth = Literal["ok", "conflict", "ambiguous"]


class _ReadModel(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True, frozen=True)


class LedgerTimelineContext(_ReadModel):
    mode: Literal["none", "single", "multiple"]
    timeline_id: UUID | None = None
    timeline_name: str | None = None
    narrative_cutoff: int | None = Field(default=None, ge=0)


class LedgerEntityReference(_ReadModel):
    entity_type: Literal[
        "character",
        "character_instance",
        "relationship",
        "storyline",
        "foreshadow",
    ]
    entity_id: UUID
    label: str
    lifecycle_state: str | None = None
    reference_missing: bool = False


class LedgerSourceReference(_ReadModel):
    source_document_id: UUID | None = None
    document_title: str | None = None
    document_position: int | None = None
    source_revision_id: UUID | None = None
    revision_number: int | None = None
    revision_is_current: bool | None = None
    source_content_hash: str | None = None
    source_start: int | None = None
    source_end: int | None = None
    binding_state: str | None = None
    commit_batch_id: UUID | None = None
    evidence_available: bool = False


class LedgerFactItem(_ReadModel):
    id: UUID
    fact_type: str
    subject: str
    predicate: str
    object_preview: str
    object_truncated: bool
    timeline_id: UUID | None = None
    dimension: str | None = None
    event_kind: str | None = None
    story_sequence: int | None = None
    created_at: datetime
    effective_state: FactEffectiveState
    effective_reason_codes: tuple[str, ...]
    included_in_current_projection: bool
    health: FactHealth
    health_reason_codes: tuple[str, ...]
    entities: tuple[LedgerEntityReference, ...] = ()
    source: LedgerSourceReference | None = None


class LedgerEventLinkView(_ReadModel):
    id: UUID
    direction: Literal["incoming", "outgoing"]
    link_type: str
    other_fact_id: UUID
    details: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class LedgerBindingView(_ReadModel):
    id: UUID
    source_document_id: UUID
    source_revision_id: UUID
    source_content_hash: str
    validity_state: str
    proposal_item_id: UUID | None = None
    commit_batch_id: UUID | None = None
    commit_batch_state: str | None = None
    created_at: datetime


class LedgerFactDetail(_ReadModel):
    schema_version: Literal["story-ledger-fact-detail/1"] = (
        "story-ledger-fact-detail/1"
    )
    novel_id: UUID
    ledger_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    item: LedgerFactItem
    object_text: str
    details: dict[str, Any] = Field(default_factory=dict)
    story_time: dict[str, Any] | None = None
    visibility: dict[str, Any] | None = None
    lifecycle_status: str
    schema_version_of_fact: str | None = None
    event_fingerprint: str | None = None
    event_links: tuple[LedgerEventLinkView, ...] = ()
    bindings: tuple[LedgerBindingView, ...] = ()


class LedgerSummary(_ReadModel):
    schema_version: Literal["story-ledger-summary/1"] = "story-ledger-summary/1"
    novel_id: UUID
    ledger_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    filter_sha256: str
    total: int
    by_fact_type: dict[str, int]
    by_effective_state: dict[str, int]
    by_health: dict[str, int]
    review_required: int


class LedgerFactPage(_ReadModel):
    schema_version: Literal["story-ledger-page/1"] = "story-ledger-page/1"
    novel_id: UUID
    ledger_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    filter_sha256: str
    items: tuple[LedgerFactItem, ...]
    next_cursor: str | None = None


class LedgerSourceExcerpt(_ReadModel):
    schema_version: Literal["story-ledger-source/1"] = "story-ledger-source/1"
    novel_id: UUID
    fact_id: UUID
    ledger_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    available: bool
    unavailable_reason: str | None = None
    document_id: UUID | None = None
    document_title: str | None = None
    document_position: int | None = None
    revision_id: UUID | None = None
    revision_number: int | None = None
    revision_is_current: bool | None = None
    source_content_hash: str | None = None
    source_range_hash: str | None = None
    source_start: int | None = None
    source_end: int | None = None
    excerpt: str = ""
    excerpt_start: int | None = None
    excerpt_end: int | None = None
    highlight_start: int | None = None
    highlight_end: int | None = None
    truncated_before: bool = False
    truncated_after: bool = False


class LedgerFactImpactPreview(_ReadModel):
    schema_version: Literal["story-ledger-fact-impact-preview/1"] = (
        "story-ledger-fact-impact-preview/1"
    )
    novel_id: UUID
    fact_id: UUID
    preview_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    currently_in_projection: bool
    current_projection_fact_count: int
    related_event_link_count: int
    embedding_rebuild_required: bool
    commit_batch_ids: tuple[UUID, ...] = ()
    batch_fact_count: int
    batch_relationship_count: int
    correction_supported: bool
    correction_block_reason: str | None = None


class LedgerBatchImpactPreview(_ReadModel):
    schema_version: Literal["story-ledger-batch-impact-preview/1"] = (
        "story-ledger-batch-impact-preview/1"
    )
    novel_id: UUID
    batch_id: UUID
    preview_snapshot_token: str
    story_ledger_version: int
    timeline: LedgerTimelineContext
    state: str
    already_reverted: bool
    batch_fact_count: int
    batch_relationship_count: int
    facts: tuple[dict[str, Any], ...]
    relationships: tuple[dict[str, Any], ...]


__all__ = [
    "FactEffectiveState",
    "FactHealth",
    "LedgerBatchImpactPreview",
    "LedgerBindingView",
    "LedgerEntityReference",
    "LedgerEventLinkView",
    "LedgerFactDetail",
    "LedgerFactImpactPreview",
    "LedgerFactItem",
    "LedgerFactPage",
    "LedgerSourceExcerpt",
    "LedgerSourceReference",
    "LedgerSummary",
    "LedgerTimelineContext",
]
